hphp/hack/src/hh_codesynthesis/agentGraphGenerator.py

```diff
@@ -0,0 +1,266 @@
+    """Context class interact with Python and Clingo.
+    We can customize each value, range, and function to experiment on
+    different settings. Refer to the test case `test_small_agent_graph`
+    for more detail usages."""
+
+    number_of_leaves = 30
+    number_of_infra_agents = 20
+    number_of_product_agents = 60
+    infra_agent_profile = {"in_degree": [0, 10], "out_degree": [5, 100]}
+    product_agent_profile = {"in_degree": [5, 20], "out_degree": [0, 5]}
+
+    def nl(self) -> Symbol:
+        return Number(self.number_of_leaves)
+
+    def nia(self) -> Symbol:
+        return Number(self.number_of_infra_agents)
+
+    def npa(self) -> Symbol:
+        return Number(self.number_of_product_agents)
+
+    def infra_agent_out_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["out_degree"][0])
+
+    def infra_agent_out_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["out_degree"][1])
+
+    def infra_agent_in_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["in_degree"][0])
+
+    def infra_agent_in_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["in_degree"][1])
+
+    def product_agent_out_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["out_degree"][0])
+
+    def product_agent_out_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["out_degree"][1])
+
+    def product_agent_in_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["in_degree"][0])
+
+    def product_agent_in_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["in_degree"][1])
+
+    def agent_connector(
+        self,
+        source_agent: Number,
+        target_agent: Number,
+        source_level: Number,
+        target_level: Number,
+    ) -> Symbol:
+        src_agent_number = source_agent.number
+        tgt_agent_number = target_agent.number
+        src_level_number = source_level.number
+        tgt_level_number = target_level.number
+        # Since clingo doesn't support Boolean return, we are returning
+        # Number(1) for true, and Number(0) for false.
+        if (
+            src_agent_number < tgt_agent_number + 200
+            and tgt_agent_number < src_agent_number
+            and tgt_level_number < src_level_number
+        ):
+            return Number(1)
+        else:
+            return Number(0)
+
+    def agent_type_connector(
+        self, source_agent: Number, target_agent: Number
+    ) -> Symbol:
+        src_agent_number = source_agent.number
+        tgt_agent_number = target_agent.number
+        # Since clingo doesn't support Boolean return, we are returning
+        # Number(1) for true, and Number(0) for false.
+        if tgt_agent_number * 3 + 200 < src_agent_number:
```

Comment:
Where does this number come from? What does *3 + 200 do?

Reply:
It's like a hash function I created during the experiment on agent graph generation. Can define any other function to provide a set of agents for the current agent to depend on. Then the solver to choose a subset from this.

Reply:
I am thinking about a "pseudo random number generator" here. With the same seed during unit test, we should get the same output. And the user can change to a seed based on UNIX time if they want variation. And the total candidates we selected for each agent should be in O(C), e.g a constant like 100 or 200.